alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/AbstractOrientingBehavior.kt

```diff
@@ -0,0 +1,318 @@
+                    if (path != null) {
+                        break
+                    }
+                }
+            }
+            path?.toMutableList() ?: mutableListOf()
+        }
+    }
+    /**
+     * The room the pedestrian is into.
+     */
+    protected var currRoom: N1? = null
+    /*
+     * The room the pedestrian is heading to.
+     */
+    private var nextRoom: N1? = null
+    /*
+     * The position the pedestrian is moving towards, no obstacle is placed
+     * between the agent and this position.
+     */
+    private var subdestination: P? = null
+    /**
+     * The edge (or better, crossing) the pedestrian is moving to.
+     */
+    protected var targetEdge: E1? = null
+    /**
+     */
+    protected enum class State {
+        /**
+         */
+        START,
+        /**
+         */
+        NEW_ROOM,
+        /**
+         */
+        MOVING_TO_DOOR,
+        /**
+         */
+        CROSSING_DOOR,
+        /**
+         */
+        MOVING_TO_FINAL,
+        /**
+         */
+        ARRIVED
+    }
+    /**
+     * This behavior is organised as a finite state machine
+     */
+    protected var state: State = State.START
+    /**
+     * Collection of areas which are found to be impasses (and thus
+     * will be avoided).
+     */
+    private val impasses: MutableSet<N1> = mutableSetOf()
+
+    /**
+     */
+    override fun updateInternalStatus(curTime: Time?, executed: Boolean, env: Environment<T, *>?) {}
+
+    /**
+     */
+    override fun getRate(): Double = timeDistribution.rate
+
+    /**
+     */
+    override fun execute() {
+        var p = env.getPosition(pedestrian)
+        when (state) {
+            State.START -> {
+                currRoom = envGraph.nodeContaining(p)
+                if (currRoom != null) {
+                    state = State.NEW_ROOM
+                }
+                /*
+                 * If the pedestrian cannot locate itself inside any room (unusual
+                 * condition), it tries to reach the closest door/passage in order
+                 * to enter one. If this isn't possible, it simply won't move.
+                 */
+                else {
+                    val closestDoor = envGraph.nodes()
+                        .flatMap { envGraph.edgesFrom(it) }
+                        .map { it to computeSubdestination(it) }
+                        .minBy { it.second.getDistanceTo(p) }
+                    if (closestDoor != null) {
+                        nextRoom = closestDoor.first.to
+                        subdestination = closestDoor.second
+                        targetEdge = closestDoor.first
+                        state = State.MOVING_TO_DOOR
+                    } else {
+                        state = State.ARRIVED
+                    }
+                }
+            }
+            State.NEW_ROOM -> {
+                if (nextRoom != null) {
+                    currRoom = if (nextRoom!!.contains(p)) {
```

Comment:
Using `!!` is ugly, and usually points to an issue. Why is `nextRoom` optional? If it really needs to be optional, then you must check for its state and fail clearly. If it's not really optional, but there are initialization issues, consider `lateinit`.